RELEASE.md

```diff
@@ -1,66 +1,78 @@
-# Releasing Roles Binaries and Publishing
+# Versioning
 
-The github binary releases of the roles and the publishing of the SRI crates are both handled in the `release.yaml` workflow. This workflow must be manually
-started by navigating to the "Actions" tab in the SRI repo, then navigating to the Release workflow section, and clicking "Run Workflow".
+SRI releases follow [SemVer 2.0.0](https://semver.org/).
 
-## Crates Publishing
+Given a version number `MAJOR.MINOR.PATCH`, we increment the:
+- `MAJOR` version when incompatible API changes are introduced
+- `MINOR` version when functionality is added in a backward compatible manner
+- `PATCH` version when backward compatible bug fixes are introduced
 
-For publishing crates we use the `cargo-release` tools because it easily handles the order in which crates can be published. Crates are published 
-in the `crates_publish` job. Currently, crates.io rate limits the publishing of more than 5 NEW crates. This means that if we have more than 5 crates
-to publish, that are not currently published to crates.io, this workflow will not publish anything. To bypass this we will need to add the `--exclude [crate_name]`
-flag for as many crates as needed to the `cargo-release` command in the "Publish" step. For example, if there are 7 new crates to publish, we will need to add 
-`--exclude [crate_1] --exclude [crate_2]` to the command. This is not an issue for publishing version updates since the crates are not NEW publishes.
+SRI has a global version, which uses git tags and keeps track of how the codebase evolves across time as a whole.
 
-### Current Successful Publishes
+All the other internal SRI crates also follow SemVer 2.0.0, but each crate version is independent of the global release version.
 
-Due to either github dependencies or a crate failing the build stage during publish not all crates are being published.
+Whenever a `PATCH` is introduced, it is applied to all the latest `MAJOR` releases.
+For example: imagine there's releases `v1.0.0`, `v1.1.0`, and `v2.0.0`. A bug is found, dating back all the way to `v1.0.0`.
+A patch is applied such that the following new tags are introduced: `v1.1.1` and `v2.0.1`.
```

Comment:
How do we do it? I guess that we should have a branch (or a tag?) for all the last x majior version of each libs.
This seems complicate. Another way could be to create the branch only when needed. We fix a bug for `lib1-2.0.0` we find that also `lib1-1.1.0` is affected and we want to backport the fix. We create a new branch from whatever was the state of the repo when lib1 have been go from v1 ti v2 and we apply the fix. If we do that only for last majior version, only when we have a bug that affect also the last majior version I do not think that we will end up with 'too many' branches, considering that most library in this repo will do not get a new majior version in many years. If this is a concern we can just say that we support the last majior version and that is not safe to use any version that is not the majior, or we can think about it when we will have more then one majior version (0 wont be maintained cause is unstable). Maybe we can just leave the various possibility here, for now we can say that we support only last majior so we do not make promise and when we will have more majior  we will think about it.

Reply:
On lines [10 and 12](https://github.com/plebhash/stratum/blob/c4399a92b9e68af72d1954e97a1c40a2430d49c9/RELEASE.md?plain=1#L10-L12) of my proposal, there is a very important distiction we need to keep in mind:

```
SRI has a global version, which uses git tags and keeps track of how the codebase evolves across time as a whole.

Every internal SRI crate also follows SemVer 2.0.0, but each crate version is independent of the global release version.
```

Let's imagine this scenario, where both `crate-A` and `crate-B` are affected by a bug on the following versions:

| Global Release Version | v2.0.0 | v1.1.0 |
|------------------------|--------|--------|
| crate-A                | v1.0.0 | v0.9.0 |
| crate-B                | v2.0.0 | v2.0.0 |

Here, Global Release Versions would be indicated as git tags.
Crate versions however, would only be indicated by `Cargo.toml` and `crates.io`. Otherwise, we would end up with too many tags.

After we patch the bug, we would have the following:
| Global Release Version (new git tags) | v2.0.1 | v1.1.1 |
|------------------------|--------|--------|
| crate-A (only bump `Cargo.toml`               | v1.0.1 | v0.9.1 |
| crate-B (only bump `Cargo.toml`)               | v2.0.1 | v2.0.1 |

Reply:
Ok make sense. Only thing I wouldn't maintain version 0 and 1. I would just bump majior version of everyting to 2. And start from there 2 reasons here:
1. We didn't followed semver so if you assume semver for what we have now you risk to end up with not compatible libraries.
2. There several 0 library that now are 1 that at 0 were very very experimental and it do not make sense to to support them

An example of that is the noise crate.

Also is startum v2 so it can make sense to start from 2. Otherwise we should remove the old library from crates.io but I dont' know if is something possible or not.

Reply:
I agree with what you're saying @Fi3, the only thing that bugs me is starting with 2.0.0, afaik we didn't really have an official release yet, and in public we always referred to things as MVP and later update. 1.0.0 is really a nice clean milestone, I am worried 2.0.0 may communicate certain level of maturity and stability which I am not sure we have yet.

> Otherwise we should remove the old library from crates.io but I don't know if is something possible or not.

Doesn't sound like something impossible, I doubt there can be any consequences to this?

Reply:
With regards to `2.0.0`, we should keep the **SRI vs SV2** distinction in mind.

Sure, SV2 implies v2, but only for the Specifications. 

That doesn't necessarily mean that as a Rust codebase, SRI should start at v2.

Reply:
With regards to crates.io removal, I found this:

https://github.com/rust-lang/crates.io/pull/1506#issuecomment-429102810

which is discouraging... it basically means we will not be able to remove the old libraries from crates.io, and we cannot "start from scratch".

I did understand your point on starting on `2.0.0` @Fi3 . It would be a way to put all SRI crates at the same starting point. But I still think starting on `2.0.0` would be bad, for the reasons I explained in my previous comment.

I wonder if we could do the same, but with all crates starting at `1.0.0`?
I did a quick search on all workspaces and I didn't see any crate that is above `1.0.0`.

Reply:
from @Fi3 (over discord):

> main issue is noise crate already 1.1.0 and is not semver 1.0.0 is not compatible with 1.1.0 but jakub told me in the call that you can mark a crate on crates.io as not usable so this could be a solution we could mark all crates like this one as not usable and in that case start from 1.1.0 and mark 1.0.0 as not usable

so assuming we will set all crates to `v1.0.0` (except `noise-sv2`), I crated this PR: https://github.com/stratum-mining/stratum/pull/785

Reply:
Sounds good. All the 0s will become 1.0.0. All the one already to 1 are yanked and we start from the very next minor. And we say that 0 is not supported anymore.